File: app/src/main/java/com/lipeng/mygithub/base/BaseAdapter.java
package com.lipeng.mygithub.base;

import android.view.View;
import android.view.ViewGroup;

import java.util.ArrayList;
import java.util.List;

/**
 * adapter基类
 * @author lipeng
 * @date 2018/1/8
 */

public abstract class BaseAdapter<T> extends android.widget.BaseAdapter {
    protected final ArrayList<T> mDataList = new ArrayList<>();

    public List<T> getDataList() {
        return mDataList;
    }

    public BaseAdapter(){

    }

    /**
     * 添加数据
     * */
    public void appendToList(List<T> dataList){
        if (null != dataList && !dataList.isEmpty()){
            mDataList.ensureCapacity(this.mDataList.size() + dataList.size());
            mDataList.addAll(dataList);
            notifyDataSetChanged();
        }
    }

    /**
     * 刷新列表
     * */
    public void refresh(List<T> list){
        if (null != list && !list.isEmpty()){
            mDataList.clear();
            mDataList.ensureCapacity(list.size());
            mDataList.addAll(list);
            notifyDataSetChanged();
        }
    }

    @Override
    public int getCount() {
        return mDataList.size();
    }

    @Override
    public Object getItem(int position) {
        if (position < this.mDataList.size()){
            return this.mDataList.get(position);
        }
        return null;
    }

    @Override
    public long getItemId(int position) {
        return (long)position;
    }

    @Override
    public View getView(int position, View convertView, ViewGroup parent) {
        return getExView(position, convertView, parent);
    }

    /**
     * 自行实现具体视图内容
     * @param position 列表的子项在数据集中的位置
     * @param convertView 复用的视图
     * @param parent 父布局
     * @return 返回构建后的视图
     * */
    protected abstract View getExView(int position, View convertView, ViewGroup parent);
}

File: app/src/main/java/com/lipeng/mygithub/base/BaseAdapter.kt
package com.lipeng.mygithub.base

import android.content.Context
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import com.lipeng.mygithub.util.ListUtils

/**
 * @author big insect
 */
abstract class BaseAdapter<VH :RecyclerView.ViewHolder, D: Any>(protected var context: Context)
    : RecyclerView.Adapter<VH>(), View.OnClickListener, View.OnLongClickListener {
    protected var dataList:ArrayList<D> = ArrayList()
    private var mClickListener :OnItemClickListener? = null
    private var mLongClickListener :OnItemLongClickListener? = null

    override fun onCreateViewHolder(parent: ViewGroup?, viewType: Int): VH {
        val view: View = LayoutInflater.from(parent?.context)
                .inflate(getLayoutId(), parent, false)
        return getViewHolder(view)
    }

    override fun onBindViewHolder(holder: VH, position: Int) {
        if (null != mClickListener){
            holder.itemView.setOnClickListener(this)
        }
        if (null != mLongClickListener) {
            holder.itemView.setOnLongClickListener(this)
        }
    }

    fun setOnItemClickListener(listener: OnItemClickListener?){
        this.mClickListener = listener
    }

    fun setOnItemLongClickListener(listener: OnItemLongClickListener?){
        this.mLongClickListener = listener
    }

    override fun onClick(v: View) {
        mClickListener?.onItemClick(v.tag)
    }

    override fun onLongClick(v: View?): Boolean {
        return mLongClickListener!!.onItemLongClick(v?.tag)
    }

    override fun getItemCount(): Int {
        return ListUtils.getSize(dataList)
    }

    override fun getItemId(position: Int): Long {
        return position.toLong()
    }
    /**
     * 获取到数据，通知视图更新
     * */
    fun addData(dataList:ArrayList<D>){
        this.dataList.clear()
        this.dataList.addAll(dataList)
        notifyDataSetChanged()
    }

    /**
     * 获取布局ID
     * */
    abstract fun getLayoutId():Int

    /**
     * 构建ViewHolder
     * */
    abstract fun getViewHolder(itemView:View):VH

    interface OnItemClickListener{
        fun onItemClick(tag :Any)
    }

    interface OnItemLongClickListener{
        /**
         * @return true 只执行此方法中的代码
         * @return false 继续响应其他监听中的事件
         * */
        fun onItemLongClick(tag: Any?):Boolean
    }
}